fs-repo-12-to-13/migration/migration.go

```diff
@@ -0,0 +1,227 @@
+// Reversible returns false, as you can just use the backup config file.
+func (m Migration) Reversible() bool {
+	return true
+}
+
+// Apply update the config.
+func (m Migration) Apply(opts migrate.Options) error {
+	log.Verbose = opts.Verbose
+	log.Log("applying %s repo migration", m.Versions())
+
+	log.VLog("locking repo at %q", opts.Path)
+	lk, err := lock.Lock2(opts.Path)
+	if err != nil {
+		return err
+	}
+	defer lk.Close()
+
+	repo := mfsr.RepoPath(opts.Path)
+
+	log.VLog("  - verifying version is '12'")
+	if err := repo.CheckVersion("12"); err != nil {
+		return err
+	}
+
+	log.Log("> Upgrading config to new format")
+
+	path := filepath.Join(opts.Path, "config")
+	if err := convertFile(path); err != nil {
+		return err
+	}
+
+	if err := repo.WriteVersion("13"); err != nil {
+		log.Error("failed to update version file to 13")
+		return err
+	}
+
+	log.Log("updated version file")
+
+	log.Log("Migration 12 to 13 succeeded")
+	return nil
+}
+
+func (m Migration) Revert(opts migrate.Options) error {
+	log.Verbose = opts.Verbose
+	log.Log("reverting migration")
+	lk, err := lock.Lock2(opts.Path)
+	if err != nil {
+		return err
+	}
+	defer lk.Close()
+
+	repo := mfsr.RepoPath(opts.Path)
+	if err := repo.CheckVersion("13"); err != nil {
+		return err
+	}
+
+	cfg := filepath.Join(opts.Path, "config")
+	if err := os.Rename(cfg+backupSuffix, cfg); err != nil {
+		return err
+	}
+
+	if err := repo.WriteVersion("12"); err != nil {
+		return err
+	}
+	if opts.Verbose {
+		log.Log("lowered version number to 12")
+	}
+
+	return nil
+}
+
+// convertFile converts a config file from 12 version to 13
+func convertFile(path string) error {
+	in, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+
+	// make backup
+	backup, err := atomicfile.New(path+backupSuffix, 0600)
+	if err != nil {
+		return err
+	}
+	if _, err := backup.ReadFrom(in); err != nil {
+		backup.Abort()
+		return err
+	}
+	if _, err := in.Seek(0, io.SeekStart); err != nil {
+		backup.Abort()
+		return err
+	}
+
+	// Create a temp file to write the output to on success
+	out, err := atomicfile.New(path, 0600)
+	if err != nil {
+		in.Close()
+		backup.Abort()
+		return err
+	}
+
+	err = convert(in, out)
+
+	in.Close()
+
+	if err != nil {
+		// There was an error so abort writing the output and clean up temp file
+		out.Abort()
+		backup.Abort()
+	} else {
+		// Write the output and clean up temp file
+		out.Close()
+		backup.Close()
+	}
+
+	return err
+}
+
+// convert converts the config from one version to another
+func convert(in io.Reader, out io.Writer) error {
+	confMap := make(map[string]any)
+	if err := json.NewDecoder(in).Decode(&confMap); err != nil {
+		return err
+	}
+
+	// run this first to avoid having both quic and quic-v1 webtransport addresses
+	runOnAllAddressFields(confMap, multiaddrPatternReplace(false, "/quic/webtransport", "/quic-v1/webtransport", "/p2p-circuit"))
+	runOnAllAddressFields(confMap, multiaddrPatternReplace(true, "/quic", "/quic-v1", "/p2p-circuit"))
```

Comment:
Is this _replacing_ all `/quic` addresses with `/quic-v1`?
For now, we should keep both of them, so that legacy nodes can still dial us.

Reply:
That sounds correct, thank you for the explanation!